android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/WifiDirectConnect.java

```diff
@@ -0,0 +1,113 @@
+import android.net.Uri;
+import android.net.wifi.p2p.WifiP2pConfig;
+import android.net.wifi.p2p.WifiP2pInfo;
+import android.net.wifi.p2p.WifiP2pManager;
+import android.net.wifi.p2p.WifiP2pManager.ActionListener;
+import android.net.wifi.p2p.WifiP2pManager.Channel;
+import android.util.Log;
+
+/**
+ * This class is for using WifiDirect to create a WifiDirect
+ * P2P connection with the Physical Web Device.
+ */
+class WifiDirectConnect {
+  private static final String TAG = "WifiDirectConnect";
+  private static Context mContext;
+  private static WifiP2pManager mManager;
+  private static Channel mChannel;
+  private static WiFiDirectBroadcastReceiver mReceiver = null;
+  private static UrlDevice mDevice;
+
+  public static void connect(Context context, UrlDevice urlDevice) {
+    mManager = (WifiP2pManager) context.getSystemService(Context.WIFI_P2P_SERVICE);
+    mChannel = mManager.initialize(context, context.getMainLooper(), null);
+    mContext = context;
+    mDevice = urlDevice;
+    mReceiver = new WiFiDirectBroadcastReceiver(mManager, mChannel);
+    IntentFilter intentFilter = new IntentFilter();
+    intentFilter.addAction(WifiP2pManager.WIFI_P2P_CONNECTION_CHANGED_ACTION);
+    mContext.registerReceiver(mReceiver, intentFilter);
+    WifiP2pConfig config = new WifiP2pConfig();
+    config.deviceAddress = Utils.getWifiAddress(mDevice);
+    config.groupOwnerIntent = 0;
+    mManager.connect(mChannel, config, new ActionListener() {
+      @Override
+      public void onSuccess() {
+        Log.d(TAG, "connect call success");
+      }
+
+      @Override
+      public void onFailure(int reason) {
+        Log.d(TAG, "connect call fail " + reason);
+      }
+    });
+  }
+
+  public static class WiFiDirectBroadcastReceiver extends BroadcastReceiver {
+    private WifiP2pManager mManager;
+    private Channel mChannel;
+
+    /**
+     * @param mManager WifiP2pManager system service
+     * @param mChannel Wifi p2p mChannel
+     * @param activity activity associated with the receiver
+     */
+    public WiFiDirectBroadcastReceiver(WifiP2pManager mManager, Channel mChannel) {
+      super();
+      this.mManager = mManager;
+      this.mChannel = mChannel;
+    }
+
+    @Override
+    public void onReceive(Context context, Intent intent) {
+      String action = intent.getAction();
+      if (WifiP2pManager.WIFI_P2P_CONNECTION_CHANGED_ACTION.equals(action)) {
+        if (mManager == null) {
+          return;
+        }
+        NetworkInfo networkInfo = (NetworkInfo) intent
+            .getParcelableExtra(WifiP2pManager.EXTRA_NETWORK_INFO);
+        if (networkInfo.isConnected()) {
+          mManager.requestConnectionInfo(mChannel,
+              new WifiP2pManager.ConnectionInfoListener() {
+            public void onConnectionInfoAvailable(final WifiP2pInfo info) {
+              Intent intent = new Intent(Intent.ACTION_VIEW);
+              intent.setData(Uri.parse("http:/" + info.groupOwnerAddress + ":" +
```

Comment:
Does this need another / in http://?  Or is that already present in info.groupOwnerAddress?